File: api/clients.go
package api

import (
	"context"
	"net/http"
	"net/url"
	"time"
)

// ClientDetected is one discovered AI-client configuration file.
//
// Detection reports STATS ONLY — never file content. On macOS, reading
// another application's configuration triggers a privacy prompt, and a bulk
// scan that prompts a dozen times is worse than no scan at all. Content is
// read only by the single-client actions, where a prompt is expected and
// explainable.
type ClientDetected struct {
	// Client is the stable client id ("claude", "cursor", ...).
	Client string `json:"client"`
	Name   string `json:"name"`
	// Placement is where the file sits ("user", "project", ...).
	Placement string `json:"placement"`
	// Shape is the configuration format family the adapter recognised.
	Shape    string `json:"shape"`
	Path     string `json:"path"`
	Writable bool   `json:"writable"`
	Size     int64  `json:"size"`
	// Modified is the file mtime: metadata from the stat, not content.
	Modified time.Time `json:"modified"`
	Note     string    `json:"note,omitempty"`
	// Denied marks a location that exists but may not be inspected.
	// "The client is not installed" and "you may not look" call for
	// opposite user actions, so a frontend must render the two differently
	// and never fold this into "not found".
	Denied bool `json:"denied,omitempty"`
	// Remediation is the operator-facing fix for a denied location.
	Remediation string `json:"remediation,omitempty"`
}

// ClientDetectResult is the answer to Detect.
type ClientDetectResult struct {
	Found []ClientDetected `json:"found"`
	// Supported lists every client agenthub can write directly, so the
	// answer to "why is my client missing" ships with the listing.
	Supported []string `json:"supported"`
}

// GatewayEntry is the MCP server entry agenthub writes into a client
// configuration: it makes the client spawn this binary as its single MCP
// server.
type GatewayEntry struct {
	Command string   `json:"command"`
	Args    []string `json:"args"`
}

// ClientConnectRequest is the body of a connect.
type ClientConnectRequest struct {
	// Profile binds this client to a profile ("" = none).
	Profile string `json:"profile,omitempty"`
	// Path overrides the configuration file to write. It names the file
	// outright and therefore may not be combined with Placement.
	Path string `json:"path,omitempty"`
	// Placement picks between the client's own "user" and "project"
	// configuration files. Empty means user-level, which is the default
	// because the entry carries the absolute path of one machine's agenthub
	// binary and project-level files are meant to be committed. A client
	// that has no such file answers E_BAD_REQUEST rather than writing to
	// the other one.
	Placement string `json:"placement,omitempty"`
	// Bin overrides the agenthub binary written into the entry.
	Bin string `json:"bin,omitempty"`
	// DryRun previews the entry and writes nothing.
	DryRun bool `json:"dry_run,omitempty"`
}

// ClientConnection is the outcome of a connect (or its dry-run preview).
type ClientConnection struct {
	Client  string       `json:"client"`
	Profile string       `json:"profile,omitempty"`
	DryRun  bool         `json:"dry_run"`
	Entry   GatewayEntry `json:"entry"`
	Path    string       `json:"path,omitempty"`
	// Backup is the copy taken before the file was rewritten. A frontend
	// should surface it: it is the operator's undo.
	Backup string `json:"backup,omitempty"`
	// Changed is false on an idempotent re-connect (the file already said
	// exactly this). Not a failure.
	Changed bool `json:"changed,omitempty"`
}

// ClientDisconnected is the outcome of a disconnect.
type ClientDisconnected struct {
	Client string `json:"client"`
	Path   string `json:"path"`
	// Removed names the entries that were deleted. Only entries agenthub
	// itself wrote are ever removed — ownership decides, never the name
	// alone, so a hand-written server that happens to share a name survives.
	Removed []string `json:"removed"`
	Backup  string   `json:"backup,omitempty"`
}

// ClientsService detects AI clients on this machine and wires them to the
// gateway.
//
// These calls carry no expectedGeneration: a client configuration file is
// not the registry, so there is no shared document to lose a
// compare-and-swap against.
type ClientsService struct{ c *Client }

// Detect scans for known client configurations.
func (s *ClientsService) Detect(ctx context.Context) (ClientDetectResult, error) {
	var out ClientDetectResult
	err := s.c.do(ctx, http.MethodGet, "/clients", nil, nil, &out)
	return out, err
}

// Connect writes the gateway entry into one client's configuration.
//
// A configuration file this process may not read answers 403 with
// E_FORBIDDEN, deliberately distinct from the uniform 404: "not installed"
// and "you may not look" call for opposite user actions.
func (s *ClientsService) Connect(
	ctx context.Context, client string, req ClientConnectRequest,
) (ClientConnection, error) {
	var out ClientConnection
	err := s.c.do(ctx, http.MethodPost, "/clients/"+url.PathEscape(client)+"/connect", nil, req, &out)
	return out, err
}

// Disconnect removes agenthub's gateway entry from one client's
// configuration. The client's other MCP servers, if any, are left exactly as
// they were.
//
// It targets the same default file a connect writes and, only when that file
// holds nothing agenthub owns, the client's other location — so an entry
// written back when connect defaulted to the project level is still removed
// rather than reported as missing.
func (s *ClientsService) Disconnect(ctx context.Context, client string) (ClientDisconnected, error) {
	var out ClientDisconnected
	err := s.c.do(ctx, http.MethodDelete, "/clients/"+url.PathEscape(client)+"/connect", nil, nil, &out)
	return out, err
}

// ClientConnectState is the answer to "is agenthub wired into this client?".
// It has five values rather than two because the three beyond yes/no each
// send the operator somewhere different, and rendering any of them as "not
// connected" points them at a connect that will not help.
type ClientConnectState string

const (
	// ClientConnected: a location holds an entry agenthub itself wrote.
	ClientConnected ClientConnectState = "connected"
	// ClientNotConnected: every location was read, none holds one.
	ClientNotConnected ClientConnectState = "not_connected"
	// ClientConnectDenied: a configuration file may not be read.
	ClientConnectDenied ClientConnectState = "denied"
	// ClientConnectUnreadable: a file is there but agenthub refuses to
	// interpret it (unparseable, oversized).
	ClientConnectUnreadable ClientConnectState = "unreadable"
	// ClientConnectUnknown: nothing was opened — a format agenthub does not
	// parse (the TOML/YAML clients).
	ClientConnectUnknown ClientConnectState = "unknown"
)

// ClientInspectedServer is one entry in an inspected client's server map.
type ClientInspectedServer struct {
	Name      string `json:"name"`
	Transport string `json:"transport,omitempty"`
	Command   string `json:"command,omitempty"`
	URL       string `json:"url,omitempty"`
	Disabled  bool   `json:"disabled,omitempty"`
	// Owned marks agenthub's own gateway entry. Ownership decides, never
	// the name: a hand-written server called "agenthub" is not this.
	Owned bool `json:"owned"`
}

// ClientInspectedFile is one configuration file of the inspected client.
type ClientInspectedFile struct {
	Path      string `json:"path"`
	Placement string `json:"placement"`
	Exists    bool   `json:"exists"`
	// Parsed false with Exists true means nobody read it: a format agenthub
	// does not parse, or the failure in Error. Servers is then empty
	// because it was not looked at, NOT because there are none.
	Parsed    bool                    `json:"parsed"`
	Connected bool                    `json:"connected"`
	Servers   []ClientInspectedServer `json:"servers,omitempty"`
	Error     string                  `json:"error,omitempty"`
}

// ClientInspection is the answer to Inspect.
type ClientInspection struct {
	Client string `json:"client"`
	Name   string `json:"name,omitempty"`
	Shape  string `json:"shape,omitempty"`
	// State is the authority; Connected is only its boolean projection.
	State      ClientConnectState    `json:"state"`
	Connected  bool                  `json:"connected"`
	Placements []string              `json:"placements,omitempty"`
	Files      []ClientInspectedFile `json:"files"`
	// Note explains a client agenthub does not parse; Manual is the
	// fragment to add to it by hand.
	Note   string `json:"note,omitempty"`
	Manual string `json:"manual,omitempty"`
}

// Inspect opens one client's configuration files and reports what is in
// them — including whether agenthub's own entry is there, which Detect
// cannot tell you because it never opens anything.
//
// This is the call that may raise a macOS privacy prompt, which is why it
// takes one named client: a frontend asks for it per row, in response to a
// user action, rather than for everything at once.
//
// A location that could not be read does not fail the call. It comes back
// with its Error, and State says denied or unreadable — never
// ClientNotConnected, which would be a claim nobody verified.
func (s *ClientsService) Inspect(ctx context.Context, client string) (ClientInspection, error) {
	var out ClientInspection
	err := s.c.do(ctx, http.MethodGet, "/clients/"+url.PathEscape(client)+"/inspect", nil, nil, &out)
	return out, err
}

File: cmd/agenthub-gui/frontend/src/pages/clients.ts
// Clients page: detecting AI clients on this machine and wiring them to the
// gateway.
//
// Detection reports STATS ONLY — never file content. On macOS, reading
// another application's configuration triggers a privacy prompt, and a bulk
// scan that prompts a dozen times is worse than no scan at all. Content is
// read only by the single-client actions, where a prompt is expected and
// explainable.
//
// A location that exists but may not be inspected is rendered as its own
// state, never folded into "not found": "the client is not installed" and
// "you may not look" call for opposite user actions, and the daemon ships the
// remediation text for the second one.

import { hub } from "../bridge";
import { clear, el, empty, icon, pageHeader, relTime } from "../dom";
import type { Page } from "../page";
import { failureBox, noticeSlot } from "../page";
import { button, checkboxInput, confirmAction, controls, field, selectInput, textInput } from "../ui";
import type { ClientConnection, ClientDetected, ClientDetectResult } from "../types";

function connectionView(c: ClientConnection): Node {
  return el("div", { class: "kvs" }, [
    kv("Client", c.client),
    kv("Profile", c.profile || "—"),
    kv("Config file", c.path || "—"),
    kv("Command", [c.entry.command, ...(c.entry.args ?? [])].join(" ")),
    kv("Backup", c.backup || "—"),
    kv("Changed", c.dry_run ? "nothing written (dry run)" : c.changed ? "yes" : "no (already wired)"),
  ]);
}

function kv(k: string, v: string): HTMLElement {
  return el("div", { class: "kv" }, [
    el("span", { class: "k", text: k }),
    el("span", { class: "v", text: v }),
  ]);
}

export function clientsPage(): Page {
  let root: HTMLElement | null = null;
  const slot = noticeSlot();
  const preview = el("div", {});
  let detected: ClientDetectResult | null = null;
  let target = "";

  async function connect(
    client: string,
    dryRun: boolean,
    profile: string,
    placement: string,
    path: string,
    bin: string,
  ): Promise<void> {
    try {
      const res = await hub.connectClient(client, {
        profile,
        // An explicit path names the file outright, so the daemon rejects it
        // together with a placement. The form sends one or the other.
        placement: path === "" ? placement : "",
        path,
        bin,
        dry_run: dryRun,
      });
      clear(preview);
      preview.append(
        el("h3", { text: dryRun ? `Preview for ${client}` : `Connected ${client}` }),
        connectionView(res),
      );
      slot.say(
        dryRun
          ? `Dry run for ${client}: nothing was written.`
          : res.changed
            ? `${client} now spawns agenthub as its single MCP server.`
            : `${client} already said exactly this — nothing changed.`,
      );
      if (!dryRun) await draw();
    } catch (err) {
      slot.fail(err);
    }
  }

  async function disconnect(client: string): Promise<void> {
    const ok = await confirmAction({
      title: `Disconnect ${client}?`,
      body: "agenthub's gateway entry is removed from that client's configuration.",
      consequences: [
        "Only entries agenthub itself wrote are removed — ownership decides, never the name, so a hand-written server that happens to share a name survives.",
        "The client stops reaching every downstream server through this hub until it is connected again.",
      ],
      confirmLabel: "Disconnect",
      danger: true,
    });
    if (!ok) return;
    try {
      const res = await hub.disconnectClient(client);
      slot.say(
        `${res.client}: removed ${res.removed.length > 0 ? res.removed.join(", ") : "nothing"} from ${res.path}` +
          (res.backup ? ` (backup at ${res.backup})` : ""),
      );
      await draw();
    } catch (err) {
      slot.fail(err);
    }
  }

  function connectForm(client: string): Node {
    const profile = textInput("", "bind to a profile (optional)");
    // User-level is the default because the entry carries this machine's
    // absolute agenthub path, and a project-level file is meant to be
    // committed. Which servers the client may see is a profile's job, not
    // this field's.
    const placement = selectInput(
      [
        { value: "user", label: "User — this machine's home directory" },
        { value: "project", label: "Project — the daemon's working tree" },
      ],
      "user",
    );
    const path = textInput("", "configuration file override (optional)");
    const bin = textInput("", "agenthub binary override (optional)");
    const dry = checkboxInput("Dry run — show the entry, write nothing", true);
    return el("div", { class: "panel panel-inset" }, [
      el("h3", { text: `Connect ${client}` }),
      el("div", { class: "form-inline" }, [
        field("Profile", profile),
        field("Placement", placement),
        field("Config path", path),
        field("Binary", bin),
      ]),
      dry.node,
      controls(
        button("Apply", "btn", () =>
          void connect(
            client,
            dry.box.checked,
            profile.value.trim(),
            placement.value,
            path.value.trim(),
            bin.value.trim(),
          ),
        ),
        button("Cancel", "btn btn-secondary", () => {
          target = "";
          void draw();
        }),
      ),
    ]);
  }

  function clientCard(c: ClientDetected): Node {
    const access = c.denied ? "not readable" : c.writable ? "writable" : "read-only";
    const accessClass = c.denied ? "badge-unhealthy" : c.writable ? "badge-healthy" : "badge-degraded";
    return el("article", { class: "client-card" }, [
      el("div", { class: "client-mark", text: (c.name || c.client).slice(0, 1).toUpperCase() }),
      el("div", { class: "client-card-main" }, [
        el("div", { class: "client-card-head" }, [
          el("div", {}, [
            el("strong", { class: "access-title", text: c.name || c.client }),
            el("div", { class: "muted", text: `${c.client} · ${c.placement} · ${c.shape}` }),
          ]),
          el("span", { class: `badge ${accessClass}`, text: access }),
        ]),
        el("div", { class: "client-path" }, [
          el("span", { class: "mono", text: c.path }),
          el("span", { class: "muted", text: `${c.size} bytes · ${relTime(c.modified)}` }),
        ]),
        c.denied && c.remediation
          ? el("div", { class: "client-remediation", text: c.remediation })
          : c.note
            ? el("div", { class: "muted", text: c.note })
            : null,
      ]),
      el("div", { class: "client-actions" }, [
        button("Connect…", "btn", () => {
          target = c.client;
          void draw();
        }),
        button("Disconnect", "btn btn-deny", () => void disconnect(c.client)),
      ]),
    ]);
  }

  // supportedHint renders the line under the table. It used to read
  // "Directly supported: <every id>", which named codex, continue and
  // open-webui as directly handled while their own rows carried a "read-only"
  // badge — and then told the reader that "anything else" needs wiring by
  // hand, when two of the three listed do. The read-only ones are now named
  // as such, from the same answer the badges come from.
  function supportedHint(d: ClientDetectResult): Node {
    const indirect = d.indirect ?? [];
    const parts = [`Supported: ${d.supported.join(", ") || "—"}.`];
    if (indirect.length > 0) {
      parts.push(
        `agenthub does not write these itself: ${indirect.join(", ")} — Connect says what to do instead.`,
      );
    }
    parts.push("Any other MCP client can still point at agenthub by hand.");
    return el("p", { class: "hint", text: parts.join(" ") });
  }

  async function draw(): Promise<void> {
    if (!root) return;
    let err: unknown = null;
    try {
      const answer = await hub.detectClients();
      detected = {
        found: answer.found ?? [],
        supported: answer.supported ?? [],
        indirect: answer.indirect ?? [],
      };
    } catch (e) {
      err = e;
      detected = null;
    }
    clear(root);
    root.append(
      pageHeader(
        "Clients",
        "Discover MCP-capable apps on this machine and connect each one to AgentHub.",
        button("Re-scan", "btn btn-primary", () => void draw()),
      ),
      slot.node,
      target ? connectForm(target) : el("span", {}),
      preview,
      el("div", { class: "privacy-note" }, [
        el("span", { class: "privacy-note-mark" }, [icon("privacy")]),
        el("span", {
          text: "Discovery reads file metadata only. File contents are opened only when you connect or disconnect a specific client.",
        }),
      ]),
      err
        ? failureBox(err)
        : (detected?.found ?? []).length === 0
          ? empty("No client configuration found on this machine.")
          : el("div", { class: "client-card-list" }, (detected?.found ?? []).map(clientCard)),
      detected ? supportedHint(detected) : el("span", {}),
    );
  }

  return {
    render(node) {
      root = node;
      return draw();
    },
    dispose() {
      root = null;
      target = "";
      clear(preview);
    },
  };
}
